zio-http/src/main/scala/zio/http/Body.scala

```diff
@@ -43,58 +45,83 @@ trait Body { self =>
 
   def isComplete: Boolean
 
-  def write(ctx: Ctx)(implicit trace: Trace): Task[Boolean]
 }
 
 object Body {
 
+  private[zio] sealed trait UnsafeWriteable extends Body
+
+  private[zio] sealed trait UnsafeBytes extends Body {
+    private[zio] def unsafeAsArray(implicit unsafe: Unsafe): Array[Byte]
+  }
+
   /**
    * Helper to create empty Body
    */
-  val empty: Body = new Body {
-    override def asChunk(implicit trace: Trace): Task[Chunk[Byte]]              = ZIO.succeed(Chunk.empty[Byte])
+  val empty: Body = EmptyBody
+
+  private[zio] object EmptyBody extends Body with UnsafeWriteable with UnsafeBytes {
+
+    override def asArray(implicit trace: Trace): Task[Array[Byte]] = ZIO.succeed(Array.empty)
+
+    override def asChunk(implicit trace: Trace): Task[Chunk[Byte]] = ZIO.succeed(Chunk.empty[Byte])
+
     override def asStream(implicit trace: Trace): ZStream[Any, Throwable, Byte] = ZStream.empty
-    override def write(ctx: Ctx)(implicit trace: Trace): Task[Boolean]          = ZIO.succeed(false)
     override def isComplete: Boolean                                            = true
 
     override def toString(): String = "Body.empty"
+
+    override private[zio] def unsafeAsArray(implicit unsafe: Unsafe): Array[Byte] = Array.empty[Byte]
   }
 
   /**
    * Helper to create Body from AsciiString
    */
-  def fromAsciiString(asciiString: AsciiString): Body = new Body {
-    override def isComplete: Boolean = true
+  def fromAsciiString(asciiString: AsciiString): Body = AsciiStringBody(asciiString)
+
+  private[zio] final case class AsciiStringBody(asciiString: AsciiString)
+      extends Body
+      with UnsafeWriteable
+      with UnsafeBytes {
+
+    override def asArray(implicit trace: Trace): Task[Array[Byte]] = ZIO.succeed(asciiString.array())
+    override def isComplete: Boolean                               = true
 
     override def asChunk(implicit trace: Trace): Task[Chunk[Byte]] =
       ZIO.succeed(Chunk.fromArray(asciiString.array()))
 
     override def asStream(implicit trace: Trace): ZStream[Any, Throwable, Byte] =
       ZStream.unwrap(asChunk.map(ZStream.fromChunk(_)))
 
-    override def write(ctx: Ctx)(implicit trace: Trace): Task[Boolean] =
-      ZIO.attempt(ctx.write(Unpooled.wrappedBuffer(asciiString.array())): Unit).as(false)
-
     override def toString(): String = s"Body.fromAsciiString($asciiString)"
+
+    private[zio] override def unsafeAsArray(implicit unsafe: Unsafe): Array[Byte] = asciiString.array()
+
   }
 
   /**
    * Helper to create Body from ByteBuf
    */
-  def fromByteBuf(byteBuf: => ByteBuf): Body = new Body {
+  def fromByteBuf(byteBuf: => ByteBuf): Body = new ByteBufBody(byteBuf)
+
+  private[zio] final class ByteBufBody(byteBuf: => ByteBuf) extends Body with UnsafeWriteable with UnsafeBytes {
+
+    def byteBufData = byteBuf
```

Comment:
If you don't make this a `lazy val`, then it will be recomputed every time. Which is fine if it's only computed once, but if it's computed more than once, then it could be useful to make it a `lazy val`.